text/3519-arbitrary-self-types-v2.md

```diff
@@ -0,0 +1,524 @@
+
+(Note that the existing algorithm isn't quite as simple as following the chain of `Deref<Target=T>`. In particular, `&T` and `&mut T` are considered as candidates too at each step; this RFC does not change that.)
+
+Because a blanket implementation is provided for users of the `Deref` trait and for `&T`/`&mut T`, the net behavior is similar. But this provides the opportunity for types which can't implement `Deref` to act as method receivers.
+
+Dereferencing a raw pointer usually needs `unsafe` (for good reason!) but in this case, no actual dereferencing occurs. This is used only to determine a list of method candidates; no memory access is performed and thus no `unsafe` is needed.
+
+## Object safety
+
+Receivers are object safe if they implement the (unstable) `core::ops::DispatchFromDyn` trait.
+
+As not all receivers might want to permit object safety or are unable to support it, object safety should remain being encoded in a different trait than the here proposed `Receiver` trait, likely `DispatchFromDyn`.
+
+This RFC does not propose any changes to `DispatchFromDyn`. Since `DispatchFromDyn` is unstable at the moment, object-safe receivers might be delayed until `DispatchFromDyn` is stabilized. `Receiver` is not blocked on further `DispatchFromDyn` work, since non-object-safe receivers already cover a big chunk of the use-cases.
+
+## Lifetime elision
+
+Arbitrary `self` parameters may involve lifetimes.
+
+This RFC currently proposes no changes to the standard lifetime elision rules. There are no known cases where ambiguity results. However, the opposite problem does apply: sometimes explicit lifetimes are required for methods with an arbitrary self type, even if they're not required for an equivalent free function. 
+
+```rust
+use std::ops::Receiver;
+
+struct SmartPtrByValue<T: ?Sized>(T);
+
+impl<T: ?Sized> Receiver for SmartPtrByValue<T> {
+    type Target = T;
+}
+
+struct SmartPtrByRef<'a, T: ?Sized>(&'a T);
+
+impl<'a, T: ?Sized> Receiver for SmartPtrByRef<'a, T> {
+    type Target = T;
+}
+
+struct Concrete(u32);
+
+impl Concrete {
+    // fn a(self: &SmartPtrByValue<Self>) -> &u32 { &self.0.0 } // does not compile
+    fn b<'a>(self: &'a SmartPtrByValue<Self>) -> &'a u32 { &self.0.0 }
+    // fn c(self: &SmartPtrB<Self>) -> &u32 {} // does not compile
+    fn d<'a, 'b>(self: &'a SmartPtrByRef<'b, Self>) -> &'a u32 { &self.0.0 }
+    fn e<'a>(self: &'a SmartPtrByRef<Self>) -> &'a u32 { &self.0.0 }
+    fn f<'a>(self: &'_ SmartPtrByRef<'a, Self>) -> &'a u32 { &self.0.0 }
+    fn g<'a, 'b>(self: &'a SmartPtrByRef<'b, Self>) -> &'b u32 { &self.0.0 }
+    fn h<'a>(self: SmartPtrByRef<'a, Self>) -> &'a u32 { &self.0.0 }
+    // fn i(self: SmartPtrByRef<Self>) -> &u32 { &self.0.0 } // does not compile
```

Comment:
This has 0 examples of eliding the lifetime of the returned reference that actually build. So clearly there's something missing, unless you want to imply that this is never supported (but even then you'd have to say that).

Reply:
Resolving this comment thread in favor of [this one about many of the same issues](https://github.com/rust-lang/rfcs/pull/3519#discussion_r1381792163) - in short all this lifetime elision section needs redoing.